src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -72,4 +92,203 @@ object RuleUtils {
+          Map())(spark)
+
+        val updatedOutput =
+          baseOutput.filter(attr => relation.schema.fieldNames.contains(attr.name))
+        baseRelation.copy(relation = relation, output = updatedOutput)
+    }
+    if (useBucketUnion) {
+      // if BucketSpec of the index is used to read the index data, we need to shuffle
+      // the appended data in the same way to avoid additional shuffle of index data.
+      getComplementIndexPlan(spark, index, plan, replacedPlan)
+    } else {
+      replacedPlan
+    }
+  }
+
+  /**
+   * Get complement plan for an index with appended data.
+   *
+   * This method consists of the following steps
+   * 1) Get a plan from originalPlan by replacing data location with appended data
+   * 2) On-the-fly shuffle for the appended data, using indexedColumns & numBuckets.
+   *   - Shuffle is located before Project to utilize Push-down Filters
+   *     - Shuffle => Project => Filter => Relation
+   *   - if Project filters indexedColumns, then Shuffle should be located after the node
+   *     - Project => Shuffle => Filter => Relation
+   * 3) Bucket-aware union both indexPlan and complementPlan to avoid repartitioning index data
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session
+   * @param index index used in replacement plan
+   * @param originalPlan original plan
+   * @param indexPlan replaced plan with index
+   * @return complementIndexPlan integrated plan of indexPlan and complementPlan
+   */
+  private def getComplementIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      originalPlan: LogicalPlan,
+      indexPlan: LogicalPlan): LogicalPlan = {
+    // 1) Replace the location of LogicalRelation with appended files
+    val complementIndexPlan = originalPlan transformUp {
+      case baseRelation @ LogicalRelation(
+            fsRelation @ HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
+            baseOutput,
+            _,
+            _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion
+        val updatedOutput =
+          baseOutput.filter(attr => index.schema.fieldNames.contains(attr.name))
+
+        val filesNotCovered = (location.allFiles
+          .map(f => FileInfo(f.getPath.toString, f.getLen, f.getModificationTime))
+          .toSet -- index.allSourceFiles).toSeq.map(f => new Path(f.name))
+
+        if (filesNotCovered.nonEmpty) {
+          val newLocation =
+            new InMemoryFileIndex(spark, filesNotCovered, Map(), None)
+          val newRelation =
+            fsRelation.copy(
+              location = newLocation,
+              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
+              spark)
+          baseRelation.copy(relation = newRelation, output = updatedOutput)
+        } else {
+          baseRelation
+        }
```

Comment:
instead of this can we do an early failure by 

```suggestion
        } else {
          return indexPlan
        }
```
or does it look like a hack?

Reply:
cc @imback82